content/articles/ansible-and-nix.md

```diff
@@ -0,0 +1,333 @@
+
+When using Ansible I've often seen the provisioning handled outside of the
+playbooks and when it is done with Ansible, the provisioning is in a separated
+role and the servers set up at fixed DNS names which then the application
+deployment roles connect to.
+
+In the voting test project I have shown how it is possible to provision servers
+and configure applications in the same playbook with Ansible.
+
+### Speed
+
+It takes about 3 minutes for both NixOps and Ansible to deploy the test
+project from scratch.
+
+For a deployment with no change, it takes about 12 seconds with Nix and 17 with
+Ansible.
+
+Of course this is entirely anecdotal, so don't read too much into these
+numbers.
+
+As much as NixOps being a bit faster than Ansible matches our experience, this
+is not something fundamental. It is possible to make Ansible deploy
+applications quickly.
+Also, where Ansible has `ansible-pull` to handle large scale deployments,
+NixOps could be problematic as the closures have to be copied over the network
+to each servers.
+
+### Services configuration
+
+Based on the same language as the Nix expressions, NixOS introduces the
+`module` abstraction. Modules are used to configure an entire system including
+UNIX users, systemd units or services like Apache and Postfix.
+This makes it possible to configure the full state of the system in a single
+top-level module. The `voting-server` attribute is in the `voting.nix` file of
+the test project is an example of a NixOS system configured using modules.
+
+You can see how the nginx module is used:
+
+```nix
+      services.nginx.enable = true;
+      services.nginx.config = ''
+  events { }
+  http {
+    include ${pkgs.nginx}/conf/mime.types;
+
+    server {
+      listen 80 default_server;
+      server_name _;
+
+      location /static/ {
+          alias ${app}/lib/python2.7/site-packages/voting/static/;
+      }
+
+      location / {
+        proxy_pass http://localhost:8000;
+      }
+    }
+  }
+  '';
+```
+
+With Ansible, some modules manage part of the system like UNIX users, but the
+configuration of most services is managed at the file level: configuration
+files are copied to a specific location and the service restarted.
+
+For example, this is how nginx is configured in the test project:
+
+```yaml
+    - template:
+        src=files/voting.conf.j2
+        dest=/etc/nginx/sites-enabled/default
+        owner=root
+      sudo: yes
+      notify: restart nginx
+```
+
+
+### Rollback
+
+Sometimes a deployment goes wrong and the best way forward is to take a step
+back.
+
+When doing an upgrade, NixOS sends the new closures, containing packages and
+configuration, and then atomically change the whole system by changing one
+symlink without overwriting the previous versions of the closures.
+
+This means that NixOps has the ability to revert a deployment atomically,
+quickly and safely in a single command.
+
+<blockquote><h4>Note</h4>Nix(OS) doesn't manage mutable state. If your
```

Comment:
it won't put the word "Note" in an h4 though. Are you suggesting that I should put the > for the blockquote and keep the <h4>Note</h4> in it?